### ISSUE TITLE
Backslashes in snippet code are output as double-backslashes

### ISSUE THREAD

**Issue by erjiang** [gosukiwi/Pasukon#3]
It doesn't seem easy or possible to write a rule like this that converts Visual-Basic-style double-quotes:

```
string
  | :STRING as :str
  |> 'return $.str.replaceAll("\x22\x22", "\x22")'
  ;
```

The generated JavaScript has:

`return $.str.replaceAll("\\x22\\x22", "\\x22")`

instead of the intended

`return $.str.replaceAll("\x22\x22", "\x22")`

The following patch seems to fix it:
```diff

--- a/lib/cli/index.js
+++ b/lib/cli/index.js
@@ -31,7 +31,7 @@ module.exports = class CLI {
     const stack = []
     parts.forEach((part) => {
       stack.push(part)
-      if (snippets.length > 0) stack.push(`"code":function($1,$2,$,$ctx){${snippets.shift()}}`)
+      if (snippets.length > 0) stack.push(`"code":function($1,$2,$,$ctx){${snippets.shift().replace(/\\\\/g, '\\')}}`)
     })
     return stack.join('')
   }
```

although it's not clear to me why `inlineFunctions` is using regexes to find code snippets.

**Comment by gosukiwi** [comment]
Good catch, if you make a PR I can merge that change, given you add tests too :)

What I want to end up with is just a string I can write to a file. So what I do is generate a JSON string, then replace part of that JSON (eg: replace `"code": "<snippet>"` with `"code": function(){<snippet>}`). The code looks like this:

```javascript
let result = JSON.stringify(parse(grammar)) // result is a string
result = this.inlineFunctions(result)
```

I do it after using `stringify` because it ignores non-primitive values:

```javascript
JSON.stringify({a: 1})
> "{"a":1}"
JSON.stringify({a: function () {}})
> "{}"
```

**Comment by erjiang** [comment]
It's ironic to get data from a parse tree by serializing it and then running a regex over it, isn't it? You will run into issues like this where the serialized string has escaped characters that shouldn't be escaped, so then you need to unescape them.

**Comment by gosukiwi** [comment]
This data is used as a cache, so it can be loaded up as a `js` file, so it makes sense to be serialized I think, unless there's something I'm not understanding.

But yes, escaping might cause issues.

**Comment by erjiang** [comment]
The alternative I was thinking of would be to use a custom serializer that can directly output what's needed. It would be maybe a little more work but hopefully would be more robust.

**Comment by gosukiwi** [comment]
Ah yes, that sounds like a good idea :) So instread of `JSON.stringify` it would use a custom serializer. If you want to tackle it, I'd love to merge it. Otherwise, it might take some time for it to happen since I'm quite busy with other things at the moment.
